File: MNIST_BLSTM/MNIST_BLSTM.cpp
#include <fstream>
#include <iostream>
#include <omp.h>
#include <stdio.h>
#include <time.h>

#include "../Neural_Networks.h"

using namespace NNCPU;

void Read_MNIST(string training_set_images, string training_set_labels, string test_set_images, string test_set_labels, int number_training, int number_test, float **input, float **target_output) {
	ifstream file(training_set_images, ifstream::binary);

	if (file.is_open()) {
		for (int h = 0, value; h < 4; h++) {
			file.read((char*)(&value), sizeof(int));
		}
		for (int h = 0; h < number_training; h++) {
			unsigned char pixel;

			for (int j = 0; j < 28 * 28; j++) {
				file.read((char*)(&pixel), 1);
				input[h][j] = pixel / 255.0;
			}
		}
		file.close();
	}
	else {
		cerr << "[Read_MNIST], " + training_set_images + " not found" << endl;
	}

	file.open(training_set_labels, ifstream::binary);

	if (file.is_open()) {
		for (int h = 0, value; h < 2; h++) {
			file.read((char*)(&value), sizeof(int));
		}
		for (int h = 0; h < number_training; h++) {
			unsigned char label;

			file.read((char*)(&label), 1);

			for (int j = 0; j < 10; j++) {
				target_output[h][j] = (j == label);
			}
		}
		file.close();
	}
	else {
		cerr << "[Read_MNIST], " + training_set_labels + " not found" << endl;
	}

	file.open(test_set_images, ifstream::binary);

	if (file.is_open()) {
		for (int h = 0, value; h < 4; h++) {
			file.read((char*)(&value), sizeof(int));
		}
		for (int h = number_training; h < number_training + number_test; h++) {
			unsigned char pixel;

			for (int j = 0; j < 28 * 28; j++) {
				file.read((char*)(&pixel), 1);
				input[h][j] = pixel / 255.0;
			}
		}
		file.close();
	}
	else {
		cerr << "[Read_MNIST], " + test_set_images + " not found" << endl;
	}

	file.open(test_set_labels, ifstream::binary);

	if (file.is_open()) {
		for (int h = 0, value; h < 2; h++) {
			file.read((char*)(&value), sizeof(int));
		}
		for (int h = number_training; h < number_training + number_test; h++) {
			unsigned char label;

			file.read((char*)(&label), 1);

			for (int j = 0; j < 10; j++) {
				target_output[h][j] = (j == label);
			}
		}
		file.close();
	}
	else {
		cerr << "[Read_MNIST], " + test_set_labels + " not found" << endl;
	}
}

int main() {
	int batch_size		= 60;
	int number_iterations = 50;
	int number_threads;
	int number_training	= 60000;
	int number_test		= 10000;
	int time_step		= 28;

	float **input = new float*[number_training + number_test];
	float **target_output = new float*[number_training + number_test];

	double epsilon		 = 0.001;
	double decay_rate	 = 0.977;
	double learning_rate = 0.005;
	double noise		 = 0.001;

	string path;

	vector<int> number_nodes = { 784, 10 };

	vector<Layer*> layer;

	// train from scratch **************************
	Neural_Networks NN = Neural_Networks(time_step);

	layer.push_back(NN.Add(new Layer("MNIST", 784 / time_step)));
	layer.push_back(NN.Add(new Layer("BN,LSTM", 100)));			 // batch normalization, LSTM forward direction
	layer.push_back(NN.Add(new Layer("BN,LSTM,backward", 100))); // batch normalization, LSTM backward direction
	layer.push_back(NN.Add(new Layer("CE,softmax", 10)));		 // cross-entropy loss, softmax activation

	layer[1]->Connect(layer[0], "W");			// fully connected
	layer[1]->Connect(layer[1], "W,recurrent"); // fully connected
	layer[2]->Connect(layer[0], "W");			// fully connected
	layer[2]->Connect(layer[2], "W,recurrent"); // fully connected
	layer[3]->Connect(layer[1], "W");			// fully connected
	layer[3]->Connect(layer[2], "W");			// fully connected

	srand(0); NN.Initialize(0.1, 0.1);

	{
		bool *time_mask = new bool[time_step];

		for (int t = 0; t < time_step; t++) {
			time_mask[t] = (t == 0 || t == time_step - 1);
		}
		layer.back()->Set_Time_Mask(time_mask);
	}
	// *********************************************

	// load pretrained model
	// Neural_Networks NN = Neural_Networks("MNIST_BLSTM.txt");

	cout << "The number of threads : ";
	cin >> number_threads;
	cin.ignore();

	omp_set_num_threads(number_threads);

	cout << "path where MNIST handwritten digits dataset is : ";
	getline(cin, path);

	for (int h = 0; h < number_training + number_test; h++) {
		input[h] = new float[number_nodes.front()];
		target_output[h] = new float[time_step * number_nodes.back()];
	}
	Read_MNIST(path + "train-images.idx3-ubyte", path + "train-labels.idx1-ubyte", path + "t10k-images.idx3-ubyte", path + "t10k-labels.idx1-ubyte", number_training, number_test, input, target_output);

	for (int h = 0; h < number_training + number_test; h++) {
		memcpy(&target_output[h][(time_step - 1) * number_nodes.back()], target_output[h], sizeof(float) * number_nodes.back());
	}
	for (int g = 0, time = clock(); g < number_iterations; g++) {
		int score[2] = { 0, };

		float **output = new float*[batch_size];

		double loss = NN.Train(batch_size, number_training, input, target_output, learning_rate, epsilon, noise);

		for (int h = 0; h < batch_size; h++) {
			output[h] = new float[time_step * number_nodes.back()];
		}
		for (int i = 0, h = 0; i < number_training + number_test; i++) {
			if (++h == batch_size || i == number_training + number_test - 1) {
				NN.Test(h, &input[i - h + 1], output);

				for (int g = 0, argmax, index = (time_step - 1) * number_nodes.back(); g < h; g++) {
					double max = 0;

					for (int j = 0; j < number_nodes.back(); j++) {
						if (max < output[g][j] + output[g][index + j]) {
							max = output[g][argmax = j] + output[g][index + j];
						}
					}
					score[(i - h + g + 1 < number_training) ? (0) : (1)] += (int)target_output[i - h + g + 1][argmax];
				}
				h = 0;
			}
		}
		printf("."); NN.Save("NN.txt");

		printf("score: %d / %d, %d / %d  loss: %lf  step %d  %.2lf sec\n", score[0], number_training, score[1], number_test, loss, g + 1, (double)(clock() - time) / CLOCKS_PER_SEC);
		learning_rate *= decay_rate;

		for (int h = 0; h < batch_size; h++) {
			delete[] output[h];
		}
		delete[] output;
	}

	for (int h = 0; h < number_training + number_test; h++) {
		delete[] input[h];
		delete[] target_output[h];
	}
	delete[] input;
	delete[] target_output;

	return 0;
}

File: MNIST_BLSTM/MNIST_BLSTM.py
import numpy as np
import time
from Neural_Networks import NNCPU

def Read_MNIST(training_set_images, training_set_labels, test_set_images, test_set_labels, number_training, number_test, image, target_output):
    # training set images
    file = open(training_set_images, 'rb')

    for _ in range(4):
        file.read(4)
        
    for h in range(number_training):
        for j in range(28 * 28):
            image[h][j] = int.from_bytes(file.read(1), byteorder='little') / 255
            
    file.close()

    # training set labels
    file = open(training_set_labels, 'rb')

    for _ in range(2):
        file.read(4)
        
    for h in range(number_training):
        label = int.from_bytes(file.read(1), byteorder='little')
        
        for j in range(10):
            target_output[h][j] = float(j == label)
            
    file.close()

    # test set images
    file = open(test_set_images, 'rb')

    for _ in range(4):
        file.read(4)
        
    for h in range(number_training, number_training + number_test):        
        for j in range(28 * 28):
            image[h][j] = int.from_bytes(file.read(1), byteorder='little') / 255
            
    file.close()

    # test set labels
    file = open(test_set_labels, 'rb')

    for _ in range(2):
        file.read(4)
        
    for h in range(number_training, number_training + number_test):
        label = int.from_bytes(file.read(1), byteorder='little')
        
        for j in range(10):
            target_output[h][j] = float(j == label)
            
    file.close()
    

batch_size      = 60
decay_rate      = 0.977
epsilon         = 0.001
learning_rate   = 0.005
noise           = 0.001
number_iterations = 50
number_output   = 10
number_training = 60000
number_test     = 10000
time_step       = 28

# train from scratch
NN = NNCPU.Neural_Networks(time_step)

NN.Add_Layer("MNIST", 784 // time_step)
NN.Add_Layer("BN,LSTM", 100)            # batch normalization, LSTM forward direction
NN.Add_Layer("BN,LSTM,backward", 100)   # batch normalization, LSTM backward direction
NN.Add_Layer("CE,softmax", 10)          # cross-entropy loss, softmax activation

NN.Connect(1, 0, "W")           # fully connected
NN.Connect(1, 1, "W,recurrent") # fully connected
NN.Connect(2, 0, "W")           # fully connected
NN.Connect(2, 2, "W,recurrent") # fully connected
NN.Connect(3, 1, "W")           # fully connected
NN.Connect(3, 2, "W")           # fully connected

NN.Initialize(0, 0.1, 0.1)

time_mask = np.empty(time_step, dtype='b')

for t in range(time_step):
    time_mask[t] = (t == 0 or t == time_step - 1)
    
NN.Set_Time_Mask(3, time_mask)


# or load pretrained model
# NN = Neural_Networks("MNIST_BLSTM.txt")

NN.Set_Number_Threads(int(input("The number of threads : ")))

image = np.zeros((number_training + number_test, 784), dtype='f')
target_output = np.zeros((number_training + number_test, time_step * 10), dtype='f')

path = input("path where MNIST handwritten digits dataset is : ")
Read_MNIST(path + "train-images.idx3-ubyte", path + "train-labels.idx1-ubyte", path + "t10k-images.idx3-ubyte", path + "t10k-labels.idx1-ubyte", number_training, number_test, image, target_output)

for h in range(number_training + number_test):
    index = (time_step - 1) * number_output
    
    for j in range(number_output):
        target_output[h][index + j] = target_output[h][j]

start = time.time()

for f in range(number_iterations):
    score = [0, 0]

    loss = NN.Train(batch_size, number_training, image, target_output, learning_rate, epsilon, noise)

    h = 0

    for i in range(number_training + number_test):
        h = h + 1
        
        if h == batch_size or i == number_training + number_test - 1:
            _input = np.array(image[i - h + 1])
            output = np.empty((h, time_step * number_output), dtype='f')

            for g in range(1, h):
                _input = np.append(_input, image[i - h + g + 1])

            NN.Test(h, _input, output)
            
            for g in range(h):
                index = (time_step - 1) * number_output
                output_mean = np.empty(number_output, dtype='f')
            
                for j in range(number_output):
                    output_mean[j] = (output[g][j] + output[g][index + j]) / 2
            
                score[0 if i - h + g + 1 < number_training else 1] += int(target_output[i - h + g + 1][np.argmax(output_mean)])
                
            h = 0

    print('.', end='')
    NN.Save("NN.txt")
    print('score: {} / {}, {} / {}, loss = {:.6f}, step {}, {:.2f} sec'.format(score[0], number_training, score[1], number_test, loss, f + 1, time.time() - start))
    
    learning_rate *= decay_rate
